tests/optim/test_oss_adascale.py

```diff
@@ -0,0 +1,124 @@
+    model.to("cuda")
+    model = DDP(model, device_ids=[rank])
+    if oss:
+        # For now, we can only wrap AdaScale over OSS. If we do it the other way around,
+        # AdaScale needs to take different parameter types, i.e. the parameter list, etc.
+        optim = AdaScale(OSS(model.parameters(), SGD, lr=0.1))
+    else:
+        optim = AdaScale(SGD(model.parameters(), lr=0.1))
+    if "input" in test_case:
+        # single iter
+        in_data = Tensor(test_case["input"][rank])
+        in_data = in_data.cuda()
+        out = model(in_data)
+        out.sum().backward()
+        assert np.allclose(optim.gain(), test_case["expected_gain"]), optim.gain()
+        optim.step()
+        optim.zero_grad()
+    else:
+        # multiple iters
+        for in_data in test_case["inputs"]:
+            in_data = Tensor(in_data[rank]).cuda()
+            out = model(in_data)
+            out.sum().backward()
+            optim.step()
+            optim.zero_grad()
+        assert np.allclose(optim.gain(), test_case["expected_gain"]), optim.gain()
+    if "expected_mean_weight" in test_case:
+        mean_weight = mean([model.module[i].weight.data.mean().item() for i in range(4)])
+        assert np.allclose(mean_weight, test_case["expected_mean_weight"]), mean_weight
+
+    dist.destroy_process_group()
+
+
+# IMPORTANT: make sure these test_cases values are sync'ed with the non-DDP
+# test in test_single_node_adascale.py. This way, we make sure gradient accumulation
+# works exactly like that in DDP & DDP+OSS.
+@skip_if_single_gpu
+@pytest.mark.parametrize(
+    "test_case",
+    [
+        # "input" value is a list of input tensors for rank 0 and rank 1.
+        {"input": [[1.0, 0], [0, 1.0]], "expected_gain": 2.0},
+        {"input": [[1.0, 1.0], [1.0, 1.0]], "expected_gain": 1.0000001249999846},
+        {"input": [[-1.0, 1.0], [1.0, -1.0]], "expected_gain": 2.0},
+        {"input": [[1.0, 4.0], [5.0, 0.5]], "expected_gain": 1.5022222222222221},
+        {"input": [[-0.2, 3.0], [5.0, 0.5]], "expected_gain": 1.9433267229211089},
+        # "inputs" to trigger multiple iteration tests, which make sure the
+        # smoothing factor calculation is also covered.
+        {"inputs": [[[-0.2, 3.3], [5.2, 0.7]], [[1.0, 4.0], [3.1, 0.1]]], "expected_gain": 1.744159431359284},
+    ],
+)
+def test_basic(test_case):
+    """Test adascale with DDP + OSS with trivial model"""
+    world_size = 2
+    temp_file_name = tempfile.mkstemp()[1]
+
+    mp.spawn(_test_basic_func, args=(world_size, temp_file_name, test_case, True), nprocs=world_size, join=True)
+
+
+@skip_if_single_gpu
+@pytest.mark.parametrize("oss", [True, False])
+def test_sequential(oss):
+    """Test adascale with DDP + OSS with a sequential model"""
+    world_size = 2
+    temp_file_name = tempfile.mkstemp()[1]
+
+    # Run multiple iterations, check the gain for both oss and non-oss cases.
+    test_case = {
+        "inputs": [[[1.0, 0], [0, 1.0]], [[0, 1.0], [1.0, 0]]],
```

Comment:
Can you add a comment about how this is calculated?

Reply:
good idea. done.